python/tvm/relay/op/contrib/cmsisnn.py

```diff
@@ -227,7 +227,7 @@ def binary_op_pattern(op):
 
     def check_qnn_binary_op(extract):
         """Check if multiply is supported by CMSIS-NN."""
-        return (
+        return bool(
```

Comment:
Im puzzled by this change.  Specifically I don't understand how the insertion of the bool() cast here changes the return type of the function, superficially it looks like the two subclauses are comparisons of str == str, both resulting in a class 'bool', applying 'and' to two bools will result in a bool.   So what does the cast achieve?

Ditto the same logic applies to each of the previous bool casts inserted above...

What am I missing?

Reply:
The problem originated because of checks for Conv2D function: check_qnn_conv2d()

Here is how it looks :
```
    conv2d.attrs.out_dtype == "int32"
            and conv2d.attrs.padding[2] == 0
            and conv2d.attrs.padding[3] == 0
            and conv2d_input.checked_type.dtype == "int8"
            and conv2d_weight.checked_type.dtype == "int8"
            and pattern.checked_type.dtype == "int8"
            and bias_dtype == "int32"
            and all([zp == 0 for zp in kernel_zp])
            and (not is_depthwise or bias_add is not None)
```
This check _conv2d.attrs.padding[2] == 0_ returns _IntImm_ type instead of _bool_. So, the fix could just be casting just _conv2d.attrs.padding[2]_ instead of casting around the whole complex boolean expression. Just to be safe always, I chose to go with the latter choice for all the operators which involves accessing attributes.

Reply:
Ah ok, that's curious, and presumably once one of those subclauses has an integer type, the subsequent application of and just propagates the numerical type rather than giving a bool.

Thanks for the explanation.

This explanation makes sense for the conv2d case where that padding comparison return intimm, but presumably it does not apply to several of the other cases where bool() has been inserted. For instance this qnn_binary_op case that I commented on... in which case it would probably be some what less confusing to the future read of this code if we didn't insert the unnecessary bool() cast.

For the case(s) where == return intimm instead of bool, I wonder whether this is the actual bug, or whether the actual bug is that the underlying == machinery is broken and returning an inappropriate type, it is certainly not illegal to have == return an arbitrary object in python, but there is a strong convention that __eq__ returns a bool, Im thinking specifically from the perspective of [the principle of least astonishment ](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

If this really is the bug (and == really should be returning a non bool object), then I'd suggest that from the readers perspective, inserting bool cast just around the subexpression that need to be cast rather than around entire subexpression is less confusing.  A comment as to why the spurious bool() cast is required might also be helpful.....

Reply:
Tagging @leandron  if he knows something about the conv2d.attrs.padding[2] == 0 returning IntImm instead of bool. Its definitely surprising. 

I'd update the code to just cast the subexpression as that makes code more readable, as you've suggested. Thanks @mshawcroft

Reply:
Where do you see the IntImm?

I have seen python side bools become IntImm s when acessed from C++ via TVM's FFIs. Is it the case here?